envoy/npds.proto

```diff
@@ -0,0 +1,114 @@
+syntax = "proto3";
+
+package pb.cilium;
+// package envoy.api.v2;
+
+import "api/address.proto";
+import "api/discovery.proto";
+import "api/rds.proto";
+
+import "google/api/annotations.proto";
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Network policy management and NPDS]
+
+// Each resource name is a network policy identifier in decimal, e.g. `123`.
+service NetworkPolicyDiscoveryService {
+  rpc StreamNetworkPolicies(stream envoy.api.v2.DiscoveryRequest)
+      returns (stream envoy.api.v2.DiscoveryResponse) {
+  }
+
+  rpc FetchNetworkPolicies(envoy.api.v2.DiscoveryRequest) returns (envoy.api.v2.DiscoveryResponse) {
+    option (google.api.http) = {
+      post: "/v2/discovery:network_policies"
+      body: "*"
+    };
+  }
+}
+
+// A network policy to be enforced on the network flows to/from associated hosts.
+message NetworkPolicy {
+  // The unique identifier of this network policy.
+  // Required.
+  uint64 policy = 1;
+
+  // The part of the policy to be enforced at ingress.
+  // Optional.
+  DirectionNetworkPolicy ingress = 2;
+
+  // The part of the policy to be enforced at egress.
+  // Optional.
+  DirectionNetworkPolicy egress = 3;
+}
+
+// A network policy to be enforced on flows either to or from hosts.
+message DirectionNetworkPolicy {
+  // The set of per-port network policies, one per destination L4 port.
+  // Every PortNetworkPolicy element in this set has a unique toPort.
+  // Required and may not be empty.
+  repeated PortNetworkPolicy per_port_policies = 1 [(validate.rules).repeated .min_items = 1];
+}
+
+// A network policy to be enforced on flows to a specific destination L4 port.
+message PortNetworkPolicy {
+  // The flows' destination L4 port number, as an unsigned 16-bit integer.
+  // Required.
+  uint32 port = 1 [(validate.rules).uint32.lte = 65535];
+
+  // The flows' L4 transport protocol.
+  // Required.
+  envoy.api.v2.SocketAddress.Protocol protocol = 2;
+  // TODO: Fix protoc-gen-validate. Adding any validate.rules to protocol above results in generating broken C++ code.
+
+  // The L7 network policy rules to be enforced on the flows to the port.
+  // A flow is allowed if any rule matches it. If no rule matches the flow, it is denied.
+  // Required and may not be empty.
+  repeated L7NetworkPolicyRule l7_rules = 3 [(validate.rules).repeated .min_items = 1];
+}
+
+// An L7 network policy rule, as a conjunction of predicates on L7 flows.
+// If all the predicates of a rule match a flow, the flow is allowed.
+message L7NetworkPolicyRule {
+  // The set of identifiers of policies of remote hosts that are allowed by this rule.
+  // This predicate matches a flow if the identifier of the policy applied on the flow's remote
+  // host is contained in this set. Required and may not be empty.
+  repeated uint64 remote_policies = 1 [(validate.rules).repeated = {min_items: 1, unique: true}];
+
+  oneof l7_rules {
```

Comment:
Would it be meaningful to allow leaving the L7 part of the policy emtpy, i.e., allow-all L7 on this port from the matching ```remote_policies```? I see this is possible for HTTP with an empty HttpNetworkPolicyRule, but that would still require the L7 protocol be HTTP, right?

Reply:
A related question: How are we expressing a visibility only rule? Or is it not related at all?

Reply:
It would have no L7 rule.

Reply:
In the current syntax L7 rule must exist, but the header matches may be empty. This what you mean?

Reply:
Right.